// @clerk/nextjs/dist/middleware/utils/authenticateRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.authenticateRequest = void 0;
const server_1 = require("../../server");
/**
 * @internal
 */
async function authenticateRequest(ctx, opts = {}) {
    const { headers, cookies } = ctx.req;
    const cookieToken = cookies['__session'];
    const headerToken = headers.authorization?.replace('Bearer ', '');
    return server_1.clerkClient.authenticateRequest({
        ...opts,
        apiKey: server_1.API_KEY,
        secretKey: server_1.SECRET_KEY,
        frontendApi: server_1.FRONTEND_API,
        publishableKey: server_1.PUBLISHABLE_KEY,
        cookieToken,
        headerToken,
        clientUat: cookies['__client_uat'],
        origin: headers.origin,
        host: headers.host,
        forwardedPort: headers['x-forwarded-port'],
        forwardedHost: headers['x-forwarded-host'],
        referrer: headers.referer,
        userAgent: headers['user-agent'],
        proxyUrl: server_1.PROXY_URL,
    });
}
exports.authenticateRequest = authenticateRequest;
