// @clerk/nextjs/dist/api/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clerkClient = exports.createClerkClient = exports.allowlistIdentifiers = exports.clients = exports.organizations = exports.invitations = exports.emails = exports.phoneNumbers = exports.emailAddresses = exports.sessions = exports.smsMessages = exports.users = exports.withAuth = exports.requireAuth = void 0;
const tslib_1 = require("tslib");
const backend_1 = require("@clerk/backend");
const clerk_1 = require("../server/clerk");
Object.defineProperty(exports, "clerkClient", { enumerable: true, get: function () { return clerk_1.clerkClient; } });
tslib_1.__exportStar(require("@clerk/backend"), exports);
const createClerkClient = backend_1.Clerk;
exports.createClerkClient = createClerkClient;
var requireAuth_1 = require("./requireAuth");
Object.defineProperty(exports, "requireAuth", { enumerable: true, get: function () { return requireAuth_1.requireAuth; } });
var withAuth_1 = require("./withAuth");
Object.defineProperty(exports, "withAuth", { enumerable: true, get: function () { return withAuth_1.withAuth; } });
const { users, smsMessages, sessions, emailAddresses, phoneNumbers, emails, invitations, organizations, clients, allowlistIdentifiers, } = clerk_1.clerkClient;
exports.users = users;
exports.smsMessages = smsMessages;
exports.sessions = sessions;
exports.emailAddresses = emailAddresses;
exports.phoneNumbers = phoneNumbers;
exports.emails = emails;
exports.invitations = invitations;
exports.organizations = organizations;
exports.clients = clients;
exports.allowlistIdentifiers = allowlistIdentifiers;
